From d52901f69e9b7952b33b219ec197308a1a20bda9 Mon Sep 17 00:00:00 2001
From: progier389 <progier@redhat.com>
Date: Fri, 13 Feb 2026 15:13:05 +0100
Subject: [PATCH] Issue 7252 - PQC - Need to iterate on SECOidTag instead of
 using OID (#7254)

* Issue 7252 - PQC - Need to iterate on SECOidTag instead of using OID

Need to dynamically iterate on SECOidTag instead of using SEC_OID_ML_DSA_* OIDs to avoid issue with upcoming nss versions and fix a RHEL build break with nss 3.112

Issue: #7252

Reviewed by: @mreynolds389, @droideck, @vashirov


* Update ldap/servers/slapd/ssl.c

Co-authored-by: Simon Pichugin <spichugi@redhat.com>

---------

Co-authored-by: Simon Pichugin <spichugi@redhat.com>
---
 ldap/servers/slapd/ssl.c | 24 +++++++++---------------
 1 file changed, 9 insertions(+), 15 deletions(-)

diff --git a/ldap/servers/slapd/ssl.c b/ldap/servers/slapd/ssl.c
index 7d5db2cdd..d05c64fb1 100644
--- a/ldap/servers/slapd/ssl.c
+++ b/ldap/servers/slapd/ssl.c
@@ -732,31 +732,25 @@ SSLPLCY_Install(void)
 {
 
     SECStatus s = 0;
-#ifdef MAX_ML_DSA_PRIVATE_KEY_LEN
     int flags = NSS_USE_ALG_IN_SIGNATURE | NSS_USE_ALG_IN_SSL;
-    static const SECOidTag oids[] = {
-        SEC_OID_ML_DSA_44,
-        SEC_OID_ML_DSA_65,
-        SEC_OID_ML_DSA_87,
-    };
-#endif
+    SECOidData *oid = NULL;
 
     s = NSS_SetDomesticPolicy();
 
-#ifdef MAX_ML_DSA_PRIVATE_KEY_LEN
     /* Should rely on the crypto module policy in FIPS mode */
     if (!slapd_pk11_isFIPS()) {
         /* Set explicitly PQC algorithm policy if it is not set by default */
-        for (size_t i=0; s == SECSuccess && i < PR_ARRAY_SIZE(oids); i++) {
-            PRUint32 oflags = 0;
-            (void) NSS_GetAlgorithmPolicy(oids[i], &oflags);
-            if ((oflags & flags) != flags) {
-                s = NSS_SetAlgorithmPolicy(oids[i], flags, 0);
+        for (SECOidTag tag = 1; s == SECSuccess && (oid = SECOID_FindOIDByTag(tag)) != NULL; tag++) {
+            if (oid->mechanism != CKM_INVALID_MECHANISM &&
+                PL_strncasecmp(oid->desc, "ML-DSA-", 7) == 0) {
+                PRUint32 oflags = 0;
+                (void) NSS_GetAlgorithmPolicy(tag, &oflags);
+                if ((oflags & flags) != flags) {
+                    s = NSS_SetAlgorithmPolicy(tag, flags, 0);
+                }
             }
         }
     }
-#endif
-
     return s ? PR_FAILURE : PR_SUCCESS;
 }
 
-- 
2.52.0

